src/credentials/CHIPCertToX509.cpp

```
@@ -0,0 +1,853 @@
+            asn1Tag = kASN1UniversalTag_UTF8String;
+        else if (attrOID == kOID_AttributeType_DomainComponent)
+            asn1Tag = kASN1UniversalTag_IA5String;
+        else
+            asn1Tag = (tlvTagNum & 0x80) ? kASN1UniversalTag_PrintableString : kASN1UniversalTag_UTF8String;
+
+        // Save the string value in the caller's DN structure.
+        dn.AttrValue.String.Value = asn1AttrVal;
+        dn.AttrValue.String.Len   = asn1AttrValLen;
+    }
+
+    // Verify that there are no further elements in the DN.
+    err = reader.VerifyEndOfContainer();
+    SuccessOrExit(err);
+
+    err = reader.ExitContainer(outerContainer);
+    SuccessOrExit(err);
+
+    // Write the ASN.1 representation of the DN...
+
+    // RDNSequence ::= SEQUENCE OF RelativeDistinguishedName
+    ASN1_START_SEQUENCE
+    {
+        // RelativeDistinguishedName ::= SET SIZE (1..MAX) OF AttributeTypeAndValue
+        ASN1_START_SET
+        {
+            // AttributeTypeAndValue ::= SEQUENCE
+            ASN1_START_SEQUENCE
+            {
+                // type AttributeType
+                // AttributeType ::= OBJECT IDENTIFIER
+                ASN1_ENCODE_OBJECT_ID(attrOID);
+
+                // value AttributeValue
+                // AttributeValue ::= ANY -- DEFINED BY AttributeType
+                err = writer.PutString(asn1Tag, Uint8::to_const_char(asn1AttrVal), asn1AttrValLen);
+                SuccessOrExit(err);
+            }
+            ASN1_END_SEQUENCE;
+        }
+        ASN1_END_SET;
+    }
+    ASN1_END_SEQUENCE;
+
+exit:
+    return err;
+}
+
+static CHIP_ERROR DecodeConvertValidity(TLVReader & reader, ASN1Writer & writer, ChipCertificateData & certData)
+{
+    CHIP_ERROR err;
+    ASN1UniversalTime asn1Time;
+    uint32_t packedTime;
+
+    ASN1_START_SEQUENCE
+    {
+        err = reader.Next(kTLVType_UnsignedInteger, ContextTag(kTag_NotBefore));
+        SuccessOrExit(err);
+        err = reader.Get(packedTime);
```

If the actual value in the TLV does not fit into a 32-bit unsigned int, this will silently drop the high bytes from it.  Should we be getting into a `uint64_t` and verifying range?  Ideally spec would define the behavior here; it does not seem to say what to do with invalid cases like this.